### PR TITLE
Fix post execution `get_pipeline()` and `pipeline.get_runs()`

### PR DESCRIPTION
**Pull request by fa9r**
## Describe changes
Getting the last run of a pipeline was broken because both `get_pipeline(...).runs` and `pipeline.get_runs()` would return the last run of the most recent pipeline version, but older versions might have newer runs.

This was very problematic because such basic behavior as the following was not working correctly:
```python
pipeline_instance.run()
run_view = pipeline_instance.get_runs()[0]  # was not guaranteed to be the run that was just executed
```

See the new integration test for a more detailed example.

To fix it, I changed the following behaviors:
- `pipeline_instance.get_runs()`: now returns the last 50 runs of this exact pipeline version (instead of the last 50 runs of the last version of the pipeline with the same name)
- `pipeline_class.get_runs()` / `get_pipeline(pipeline_name).runs ` / `get_pipeline(pipeline_class).runs`: now return the last 50 runs of all pipeline versions with the given pipeline name (instead of the last 50 runs of the last version only)

### Technical Details
- I introduced a new `PipelineClassView` post execution class that is just a namespace and has a list of all versions associated to that pipeline name. Calling `get_pipeline("pipeline_name")` or `get_pipeline(pipeline_class)` will now return such a `PipelineClassView` instead of the `PipelineView`.
- I reimplemented `BasePipeline.get_runs()` as a Descriptor that has different behavior based on whether the method is called on a pipeline class or a pipeline instance.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [x] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Comment by schustmi** [comment]
On a general note, wouldn't `PipelineView` and `PipelineVersionView` be more appropriate names?

File: src/zenml/pipelines/base_pipeline.py [hunk 1]
```diff
@@ -143,6 +143,45 @@ def __new__(
...
+        if instance is None:
+            pipeline_view = get_pipeline(cls)
+        else:
+            pipeline_view = get_pipeline(instance)
+
+        if pipeline_view:
+            return lambda: pipeline_view.runs
+        raise RuntimeError(
+            "The pipeline view for this pipeline was not found. Are you sure "
+            "this pipeline has been run already?"
```
Comment (strickvl) [inline src/zenml/pipelines/base_pipeline.py:181]: Rather than asking a question, I'd maybe rephrase this as "Please check that..."

File: src/zenml/post_execution/pipeline.py [hunk 2]
```diff
@@ -74,56 +70,51 @@ def get_pipeline(
...
+    # Pipeline instance: find the corresponding pipeline version in the DB.
+    if isinstance(pipeline, BasePipeline):
         pipeline_model = pipeline._get_registered_model()
         if pipeline_model:
             return PipelineView(model=pipeline_model)
         else:
             return None
+
+    # Otherwise, for pipeline name or class, determine the name first.
+    if isinstance(pipeline, str):
```
Comment (strickvl) [inline src/zenml/post_execution/pipeline.py:98]:
```suggestion
    elif isinstance(pipeline, str):
```

Maybe, just to preserve the logic of this conditional, even if the `return` would bypass the rest?

File: src/zenml/post_execution/pipeline.py [hunk 3]
```diff
@@ -134,11 +125,78 @@ def get_pipeline(
...
+    @property
+    def num_runs(self) -> int:
+        """Returns the number of runs of this pipeline name/class.
+
+        This is the sum of all runs of all versions of this pipeline.
+
+        Returns:
+            The number of runs of this pipeline name/class.
+        """
+        return sum([version.num_runs for version in self.versions])
```
Comment (strickvl) [inline src/zenml/post_execution/pipeline.py:163]:
```suggestion
        return sum(version.num_runs for version in self.versions)
```

Save a bit of memory with a generator expression instead of a list comprehension?

**Review by strickvl — CHANGES_REQUESTED** [review]

File: src/zenml/post_execution/pipeline.py (hunk 2 above)
Comment (fa9r) [reply src/zenml/post_execution/pipeline.py:98]: I think `if` is preferable over `elif` here since certain linters like pylint even complain if you do `elif` since the "else" part is effectively redundant

File: src/zenml/post_execution/pipeline.py [hunk 4]
```diff
@@ -134,11 +125,78 @@ def get_pipeline(
...
+
+    @property
+    def versions(self) -> List["PipelineView"]:
+        """Returns all versions/instances of this pipeline name/class.
+
+        Returns:
+            A list of all versions of this pipeline.
+        """
+        client = Client()
+        pipelines = client.list_pipelines(
```
[the remark this answers is not in the document]
Comment (fa9r) [reply src/zenml/post_execution/pipeline.py:147]: Hmm yeah probably. Pagination + post execution don't go along well in general. I'll depaginate it for now.

[the remark this answers is not in the document]
Comment (htahir1) [reply src/zenml/post_execution/pipeline.py:147]: @fa9r  Wouldnt it make sense to pass the pagination params as first class parameters in the post execution method?

[the remark this answers is not in the document]
Comment (AlexejPenner) [reply src/zenml/post_execution/pipeline.py:147]: in the long run i agree with you Hamza, but I think that takes a bit of thought and care to figure out to what extent to do this (will we include the `contains` language?). As this ticket was supposed to be a quickfix.

File: tests/integration/functional/post_execution/test_pipeline.py [hunk 5]
```diff
@@ -87,3 +93,98 @@ def test_get_pipeline_fails_for_wrong_type(clean_client, wrong_pipeline_type):
     """Test that `get_pipeline` fails for wrong input types."""
     with pytest.raises(RuntimeError):
         get_pipeline(wrong_pipeline_type)
+
+
+def test_get_runs(clean_client, connected_two_step_pipeline):
```
Comment (AlexejPenner) [inline tests/integration/functional/post_execution/test_pipeline.py:98]: Love the thorough tests!

**Review by AlexejPenner — APPROVED** [review]
:seal:

File: src/zenml/post_execution/pipeline.py [hunk 6]
```diff
@@ -144,12 +146,16 @@ def versions(self) -> List["PipelineView"]:
             A list of all versions of this pipeline.
         """
         client = Client()
-        pipelines = client.list_pipelines(
-            workspace_id=client.active_workspace.id,
-            name=self.name,
-            sort_by="desc:created",
+
+        pipelines = depaginate(
+            partial(
```
Comment (strickvl) [inline src/zenml/post_execution/pipeline.py:151]: cool

**Review by strickvl — APPROVED** [review]
🙌
